[FIX] Re-align the text and the schema for filenames in derivatives

mguaypaq opened the pull request:
I initially set out to add the `desc` entity to the filename templates for segmentations (in [the 'Segmentations' subsection](https://bids-specification.readthedocs.io/en/stable/05-derivatives/03-imaging.html#segmentations)), since the BIDS spec says [here](https://bids-specification.readthedocs.io/en/stable/05-derivatives/01-introduction.html) that:

> When necessary to distinguish two files that do not otherwise have a distinguishing entity, the [_desc-<label>](https://bids-specification.readthedocs.io/en/stable/appendices/entities.html#desc) entity SHOULD be used.

But then I noticed that this repository also contains a YAML version of the spec, and the `desc` entity already appears there. What's more, I noticed that the `atlas` filename entity has the opposite problem: it appears in the text (added in #997), but not in the YAML.

I also noticed that the `[ses-<label>/]` folder layer was missing from the filename templates in the text about Derivatives (probably because they can't use the nifty `MACROS___make_filename_template` construct), so I fixed that too.

This pull request fixes all three problems.

I'm not sure if this pull request counts as a [FIX] or a [SCHEMA]: it feels to me like a bugfix, but technically it does touch the YAML files.

(This is my first contribution, but I'll wait until it's accepted before adding myself to [the Recent Contributors page on the wiki](https://github.com/bids-standard/bids-specification/wiki/Recent-Contributors).)

tsalo commented on src/derivatives/imaging.md:
```diff
@@ -332,8 +335,9 @@ Template:
-            <source_entities>[_space-<space>][_atlas-<label>][_res-<label>][_den-<label>][_label-<label>]_probseg.nii.gz
+        [ses-<label>/]
+            func|anat|dwi/
+                <source_entities>[_space-<space>][_atlas-<label>][_res-<label>][_den-<label>][_label-<label>][_desc-<label>]_probseg.nii.gz
```
I don't think `den` is applicable to nifti files.

tsalo replied:
```suggestion
                <source_entities>[_space-<space>][_atlas-<label>][_res-<label>][_label-<label>][_desc-<label>]_probseg.nii.gz
```

tsalo commented on src/derivatives/imaging.md:
```diff
@@ -406,8 +410,9 @@ Template:
-            <source_entities>[_hemi-{L|R}][_space-<space>][_atlas-<label>][_res-<label>][_den-<label>]_dseg.{label.gii|dlabel.nii}
+        [ses-<label>/]
+            anat/
+                <source_entities>[_hemi-{L|R}][_space-<space>][_atlas-<label>][_res-<label>][_den-<label>][_desc-<label>]_dseg.{label.gii|dlabel.nii}
```
I don't think `res` is applicable to gifti files, though it _is_ applicable to cifti files.

tsalo replied:
```suggestion
                <source_entities>[_hemi-{L|R}][_space-<space>][_atlas-<label>][_den-<label>][_desc-<label>]_dseg.label.gii
                <source_entities>[_hemi-{L|R}][_space-<space>][_atlas-<label>][_res-<label>][_den-<label>][_desc-<label>]_dseg.dlabel.nii
```

mguaypaq replied:
Do you have any thoughts on how to represent in the template the fact that accompanying `.json` and `.tsv` files are allowed? Three options that spring to mind:
1. Include these file extension options on _both_ lines, that is:

    ```
    ..._dseg.{label.gii|json|tsv}
    ..._dseg.{dlabel.nii|json|tsv}
    ```

    I think this is a bit clearer, but technically redundant.
2. Include the file extension options only on the more permissive line, that is:

    ```
    ..._dseg.{label.gii}
    ..._dseg.{dlabel.nii|json|tsv}
    ```

    Maybe this would be confusing.
3. Have a separate line for each of `label.gii`, `dlabel.nii`, `json`, `tsv`, I guess forbidding `res` only for `dlabel.nii`?

tsalo commented on src/derivatives/imaging.md:
```diff
@@ -147,8 +148,9 @@ Template:
-            <source_entities>[_space-<space>][_res-<label>][_den-<label>][_label-<label>][_desc-<label>]_mask.nii.gz
+        [ses-<label>/]
+            anat|func|dwi/
+                <source_entities>[_space-<space>][_res-<label>][_den-<label>][_label-<label>][_desc-<label>]_mask.nii.gz
```
```suggestion
                <source_entities>[_space-<space>][_res-<label>][_label-<label>][_desc-<label>]_mask.nii.gz
```

I don't know if we plan to support gifti or cifti masks. If so, then this snippet will need filename templates for those as well.

mguaypaq replied:
I'd be happy to make this change, but then it also needs to be reflected in the YAML files. Before I can do that, though, a question:

I'm noticing that the `dwi` datatype also allows the file extensions `.bvec` and `.bval`. Are those relevant for masks? Or should only `.json`, `.nii` and `.nii.gz` be allowed?

tsalo replied:
I don'tthink `.bvec` and `.bval` would apply to masks. @mattcieslak WDYT?

tsalo commented on src/derivatives/imaging.md:
```diff
@@ -267,8 +269,9 @@ Template:
-            <source_entities>[_space-<space>][_atlas-<label>][_res-<label>][_den-<label>]_dseg.nii.gz
+        [ses-<label>/]
+            anat|func|dwi/
+                <source_entities>[_space-<space>][_atlas-<label>][_res-<label>][_den-<label>][_desc-<label>]_dseg.nii.gz
```
```suggestion
                <source_entities>[_space-<space>][_atlas-<label>][_res-<label>][_desc-<label>]_dseg.nii.gz
```

tsalo reviewed (dismissed):
Thank you for addressing this. I think the one lingering problem is that the set of allowed entities will depend on the extension (i.e., nifti vs. gifti vs. cifti).

I also think we're not far from being able to generate these filename templates with macros (the way we do for the raw BIDS filename templates), but your fixes are much appreciated in the meantime.

Remi-Gau commented:
@effigies @tsalo 
before @mguaypaq spends too much time on this, do we think it will be possible in the "near" future to use the schema directly to render those filename templates?

Also to estimate how thorough we want to be on this.

Remi-Gau commented:
switching to draft PR

mguaypaq commented:
This should have been a simple first contribution PR, but it languished for a year. Anyway, I believe I have addressed all the feedback, so that the text and yaml line up with each other. Could I get another round of reviews?
